SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityPendingTaskQueueProcessor.java

```diff
@@ -0,0 +1,419 @@
+        handled.add(pendingTask);
+      }
+    }
+  }
+
+  public Set<SingularityPendingTaskId> getHandledTasks() {
+    synchronized (handled) {
+      return new HashSet<>(handled);
+    }
+  }
+
+  // Method only for testing. Need a synchronous way of waiting for task launch in unit tests
+  @VisibleForTesting
+  public boolean drainHandledTasks() {
+    long timeout = 500;
+
+    long start = System.currentTimeMillis();
+    while (!handled.isEmpty()) {
+      try {
+        Thread.sleep(50);
+      } catch (InterruptedException ie) {
+        LOG.warn("Interrupted waiting for tasks to schedule");
+      }
+      if (System.currentTimeMillis() - start > timeout) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public void removePendingTask(SingularityPendingTaskId toRemove) {
+    synchronized (handled) {
+      if (!preLockQueue.remove(toRemove)) {
+        LOG.warn("Unable to remove pending task {}", toRemove.getId());
+      }
+    }
+  }
+
+  public void stop() {
+    running.set(false);
+    if (lockFuture != null) {
+      lockFuture.cancel(false);
+    }
+  }
+
+  private void runLockWait() {
+    Map<SingularityPendingTaskId, CompletableFuture<Void>> waitFutures = new HashMap<>();
+    while (running.get()) {
+      long start = System.currentTimeMillis();
+      try {
+        if (waitFutures.size() < PARALLEL_LOCK_WAIT) {
+          SingularityPendingTaskId pendingTask = preLockQueue.poll(
+            1000,
+            TimeUnit.MILLISECONDS
+          );
+          if (pendingTask == null) {
+            continue;
+          }
+          waitFutures.put(
+            pendingTask,
+            CompletableFuture.runAsync(
+              () -> waitLockAndLaunch(start, pendingTask),
+              lockWaitExecutor
+            )
+          );
+        } else {
+          Set<SingularityPendingTaskId> keys = new HashSet<>(waitFutures.keySet());
+          for (SingularityPendingTaskId key : keys) {
+            if (waitFutures.get(key).isDone()) {
+              waitFutures.remove(key);
+            }
+          }
+          try {
+            Thread.sleep(100);
+          } catch (InterruptedException ie) {
+            LOG.warn("Interrupted waiting between lock queue loops");
+            running.set(false);
+          }
+        }
+      } catch (Exception e) {
+        LOG.error("Error running task launch", e);
+      }
+    }
+  }
+
+  private void waitLockAndLaunch(long start, SingularityPendingTaskId pendingTaskId) {
+    try {
+      lockStarts.put(
+        pendingTaskId.getId(),
+        lock.lock(pendingTaskId.getRequestId(), LOCK_NAME)
+      );
+      LOG.trace(
+        "Got request lock to launch pending task {} after {}",
+        pendingTaskId.getId(),
+        System.currentTimeMillis() - start
+      );
+      // Run this on the single thread responsible for offer matching and launches
+      CompletableFuture
+        .runAsync(() -> matchAndLaunch(start, pendingTaskId), taskLaunchExecutor)
```

Comment:
Somewhat concerned about the performance of only matching/launching a single task at a time - not sure if it'd be worth modifying matchAndLaunch to pull all safely locked tasks for matching instead of explicitly passing task ids 1 at a time.